### PR TITLE
Replace CUDA-specific set_index(_from) method from DeviceGuard with set_device.

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13133 Add c10::Stream, make at::cuda::CUDAStream use it.&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10855883/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13232 Delete DeviceGuard(int64_t) constructor.&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10858024/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#13275 Replace CUDA-specific set_index(_from) method from DeviceGuard with set_device.**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D12832080/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13342 Generalize DeviceGuard to work with arbitrary DeviceType.&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D12849620/)

This resulted in a bunch of knock-on changes, which I will now
describe:

- s/original_index/original_device/
- s/last_index/last_device/
- A bunch of places that used set_index, now use CUDAGuard (which does have
  set_index) because they were CUDA-specific code.

Major caveat: DeviceGuard doesn't *actually* work non-CUDA/CPU devices, To make
that happen, I plan on totally replacing the implementation of DeviceGuard; what
I mostly care about here is wrangling the API into an acceptable state.

Differential Revision: [D12832080](https://our.internmc.facebook.com/intern/diff/D12832080/)

### CODE REVIEW

File: aten/src/ATen/DeviceGuard.h [hunk 1]
```diff
@@ -71,7 +68,7 @@ struct DeviceGuard {
     return *this;
   }
 
-  /// Resets the device to the index that was active at construction of the
+  /// Resets the device to the device that was active at construction of the
```
Comment (gchanan) [inline aten/src/ATen/DeviceGuard.h]: nit: can get rid of the first "the device"

**Review by gchanan — APPROVED** [review]
do we still have any cpu-only tests?  I didn't look at the #ifdef CUDA stuff closely.

**Comment by ezyang** [comment]
Yes, anything building with system Python and non-standard GCC still runs tests.

**Comment by dzhulgakov** [comment]
So everything specialized to CUDA would live in cuda:: namespace, right?

cc  @bddppq to double check that it makes hipify possible/easier?

**Comment by ezyang** [comment]
@dzhulgakov Yep, precisely correct.

**Pull request closed without merging**